native-gradle-plugin/src/main/java/org/graalvm/buildtools/gradle/MergeAgentFiles.java

```diff
@@ -0,0 +1,139 @@
+
+class MergeAgentFiles implements Action<Task> {
+    private final Provider<Boolean> agent;
+    private final Provider<String> graalvmHomeProvider;
+    private final Provider<Directory> outputDir;
+    private final Provider<Boolean> disableToolchainDetection;
+    private final ExecOperations execOperations;
+    private final NativeImageOptions options;
+    private final FileSystemOperations fileOperations;
+    private final Logger logger;
+
+    MergeAgentFiles(Provider<Boolean> agent,
+                    Provider<String> graalvmHomeProvider,
+                    Provider<Directory> outputDir,
+                    Provider<Boolean> disableToolchainDetection,
+                    NativeImageOptions options,
+                    ExecOperations execOperations,
+                    FileSystemOperations fileOperations,
+                    Logger logger) {
+        this.agent = agent;
+        this.graalvmHomeProvider = graalvmHomeProvider;
+        this.outputDir = outputDir;
+        this.disableToolchainDetection = disableToolchainDetection;
+        this.options = options;
+        this.execOperations = execOperations;
+        this.fileOperations = fileOperations;
+        this.logger = logger;
+    }
+
+    @Override
+    public void execute(Task task) {
+        if (agent.get()) {
+            File nativeImage = findNativeImageExecutable(options, disableToolchainDetection, graalvmHomeProvider, execOperations, GraalVMLogger.of(logger));
+            File workingDir = nativeImage.getParentFile();
+            File launcher = new File(workingDir, nativeImageConfigureFileName());
+            if (!launcher.exists()) {
+                logger.info("Installing native-image-configure");
+                execOperations.exec(spec -> {
+                    spec.executable(nativeImage);
+                    spec.args("--macro:native-image-configure-launcher");
+                });
+                NativeImageUtils.maybeCreateConfigureUtilSymlink(launcher, nativeImage.toPath());
+            }
+            if (launcher.exists()) {
+                File[] files = outputDir.get().getAsFile().listFiles();
+                List<String> args = new ArrayList<>(files.length + 2);
+                args.add("generate");
+                sessionDirectoriesFrom(files)
+                        .map(f -> "--input-dir=" + f.getAbsolutePath())
+                        .forEach(args::add);
+                if (args.size() > 1) {
+                    logger.info("Merging agent files");
+                    args.add("--output-dir=" + outputDir.get().getAsFile().getAbsolutePath());
+                    ExecResult exec = execOperations.exec(spec -> {
+                        spec.executable(launcher);
+                        spec.args(args);
+                        spec.setStandardOutput(System.out);
+                        spec.setErrorOutput(System.err);
+                    });
```

Comment:
I am not convinced we should merge them. The "merge" logic can be seen as an internal implementation detail of the agent implementation: instead of outputting _n_ files, it should output a _single_ file. Therefore, this logic needs to be invoked as part of the task execution itself, so that, for example, we don't mix files from different invocations.

The "process agent files" task, on the other hand, is, as its name states, a _post processing_ task which, yes, could be made more configurable, but I don't think we should collate both aspects.

Reply:
That sounds reasonable. I have no strong opinions about this, so exact steps in my proposal aren't set in stone. My logic was that we might want to do it in a single `native-image-configure` invocation so it sounded like a job for a single task.

I however believe that we should (somehow) utilize `native-image-configure` to filter out Gradle entries, and add a way for end user to append values into `getFilterableEntries`.

Reply:
You seem to imply that `native-image-configure` can be used to filter entries. Is this documented somewhere?

Reply:
```
$ native-image-configure --help
```
contains the following argument:
```
    --exclude-classes=<class-pattern>
                          adds a single rule to exclude classes matching the
                          specified pattern, potentially overriding rules from
                          preceding parameters. Can be specified several times.
```

Reply:
I see. Let's start with what we have now and we can always change the implementation if needed, I'm not too worried.